c/planarityTestGraphFunctionality.c

```diff
@@ -408,15 +315,120 @@ int testAllGraphs(graphP theGraph, char command, char *inputStr, testAllStatsP s
 	if (freeG6ReadIterator(&pG6ReadIterator) != OK)
 		ErrorMessage("Unable to free G6ReadIterator.\n");
 	
-	return exitCode;
+	return Result;
 }
 
-int _getNumCharsToReprInt(int theNum)
+int outputTestAllGraphsResults(char command, testAllStatsP stats, char * infileName, char *outfileName, char **outputStr)
 {
-	int numCharsRequired = 1;
+	int Result = OK;
 
-	while(theNum /= 10)
-		numCharsRequired++;
+	FILE *outputFileP;
 
-	return numCharsRequired;
-}
\ No newline at end of file
+	int charsAvailForFilename = 0;
+	char *messageFormat = NULL;
+	char messageContents[MAXLINE + 1];
+
+	strOrFileP testOutput = NULL;
+
+	if (outfileName != NULL)
+	{
+		outputFileP = fopen(outfileName, "w");
+		if (outputFileP == NULL)
+		{
+			charsAvailForFilename = (int) (MAXLINE - strlen(outfileName));
+			messageFormat = "Unable to open file \"%.*s\" for output.\n";
+			sprintf(messageContents, messageFormat, charsAvailForFilename, outfileName);
+			ErrorMessage(messageContents);
+			return NOTOK;
+		}
+		
+		testOutput = sf_New(outputFileP, NULL);
+	}
+	else
+	{
+		if ((*outputStr) == NULL)
+			(*outputStr) = (char *) malloc(1 * sizeof(char));
+		
+		if ((*outputStr) == NULL)
+		{
+			ErrorMessage("Unable to allocate memory for outputStr.\n");
+			return NOTOK;
+		}
+
+		testOutput = sf_New(NULL, (*outputStr));
+	}
+
+	if (testOutput == NULL)
+	{
+		ErrorMessage("Unable to set up string-or-file container for test output.\n");
+
+		if (outputFileP != NULL)
+		{
+			fclose(outputFileP);
+			outputFileP = NULL;
+		}
+
+		if ((*outputStr) != NULL)
+		{
+			free(outputStr);
+			outputStr = NULL;
+		}
+
+		return NOTOK;
+	}
+
+	char *finalSlash = strrchr(infileName, FILE_DELIMITER);
+	char *infileBasename = finalSlash ? (finalSlash + 1) : infileName;
+
+	char *headerFormat = "FILENAME=\"%s\" DURATION=\"%.3lf\"\n";
+	char *headerStr = (char *) malloc(
+										(
+											strlen(headerFormat) +
+											strlen(infileBasename) +
+											strlen("-1.7976931348623158e+308") + // -DBL_MAX from float.h
+											3
+										) * sizeof(char));
+	if (headerStr == NULL)
+	{
+		ErrorMessage("Unable allocate memory for output file header.\n");
+		sf_Free(&testOutput);
+		return NOTOK;
+	}
+
+	sprintf(headerStr, headerFormat, infileBasename, stats->duration);
+
+	sf_fputs(headerStr, testOutput);
+	free(headerStr);
+	headerStr = NULL;
+
+	char *resultsStr = (char *) malloc(
+										(
+											3 +_getNumCharsToReprInt(stats->numGraphsRead) +
+											1 + _getNumCharsToReprInt(stats->numOK) +
+											1 + _getNumCharsToReprInt(stats->numNONEMBEDDABLE)+
+											1 + 8 + // either ERROR or SUCCESS, so the longer of which is 7 + 1 chars
+											3
+										) * sizeof(char));
+	if (resultsStr == NULL)
+	{
+		ErrorMessage("Unable allocate memory for results string.\n");
+		sf_Free(&testOutput);
```

Comment:
This also doesn't fclose the output file or free the output string in the testOutput, but once the resultStr malloc and NULL test move up, the only thing you will have made yet is headerStr, so that would need to be freed here.